File: models/Gpt-3.5-turbo/cpp/code/problems/additional_problems/Tree_Traversals.cpp

#include <iostream>
#include <vector>
#include <unordered_map>

using namespace std;

vector<int> postorder;
unordered_map<int, int> inorderIndices;

void buildPostorder(vector<int>& preorder, int preStart, int preEnd, int inStart, int inEnd) {
    if (preStart > preEnd || inStart > inEnd) {
        return;
    }
    
    int rootValue = preorder[preStart];
    int rootIndex = inorderIndices[rootValue];
    
    buildPostorder(preorder, preStart + 1, preStart + (rootIndex - inStart), inStart, rootIndex - 1);
    buildPostorder(preorder, preStart + (rootIndex - inStart) + 1, preEnd, rootIndex + 1, inEnd);
    
    postorder.push_back(rootValue);
}

vector<int> getPostorder(vector<int>& preorder, vector<int>& inorder) {
    int n = preorder.size();
    
    // Store the indices of elements in the inorder traversal
    for (int i = 0; i < n; i++) {
        inorderIndices[inorder[i]] = i;
    }
    
    buildPostorder(preorder, 0, n - 1, 0, n - 1);
    
    return postorder;
}

int main() {
    int n;
    cin >> n;
    
    vector<int> preorder(n);
    vector<int> inorder(n);
    
    for (int i = 0; i < n; i++) {
        cin >> preorder[i];
    }
    
    for (int i = 0; i < n; i++) {
        cin >> inorder[i];
    }
    
    vector<int> result = getPostorder(preorder, inorder);
    
    for (int i = 0; i < n; i++) {
        cout << result[i] << " ";
    }
    
    cout << endl;
    
    return 0;
}

File: models/Gpt-3.5-turbo/py/code/additional_problems/Tree_Traversals.py

def build_postorder(preorder, inorder):
    if len(preorder) == 0:
        return []
    
    root = preorder[0]
    idx = inorder.index(root)
    
    left_subtree_preorder = preorder[1:idx+1]
    left_subtree_inorder = inorder[:idx]
    right_subtree_preorder = preorder[idx+1:]
    right_subtree_inorder = inorder[idx+1:]
    
    left_subtree_postorder = build_postorder(left_subtree_preorder, left_subtree_inorder)
    right_subtree_postorder = build_postorder(right_subtree_preorder, right_subtree_inorder)
    
    return left_subtree_postorder + right_subtree_postorder + [root]

# Read input
n = int(input())
preorder = list(map(int, input().split()))
inorder = list(map(int, input().split()))

# Call the function and print the result
postorder = build_postorder(preorder, inorder)
print(' '.join(map(str, postorder)))


def build_postorder(preorder, inorder):
    def build_postorder_helper(pre_start, pre_end, in_start, in_end):
        if pre_start > pre_end:
            return []
        
        root = preorder[pre_start]
        idx = index_map[root]
        
        left_subtree_size = idx - in_start
        right_subtree_size = in_end - idx
        
        left_subtree_postorder = build_postorder_helper(pre_start + 1, pre_start + left_subtree_size, in_start, idx - 1)
        right_subtree_postorder = build_postorder_helper(pre_start + left_subtree_size + 1, pre_end, idx + 1, in_end)
        
        return left_subtree_postorder + right_subtree_postorder + [root]
    
    index_map = {val: idx for idx, val in enumerate(inorder)}
    return build_postorder_helper(0, n - 1, 0, n - 1)
